src/Adapter/Product/Image/CommandHandler/SetProductImagesForAllShopHandler.php

```diff
@@ -0,0 +1,154 @@
+    {
+        $shopIdsAssociatedToProduct = array_map(
+            static function (ShopId $shopId): int {
+                return $shopId->getValue();
+            },
+            $this->productMultiShopRepository->getAssociatedShopIds($productId)
+        );
+
+        return $shopIdsAssociatedToProduct;
+    }
+
+    /**
+     * @param ProductImageSetting[] $productImageSettings
+     * @param int $imageId
+     *
+     * @return ProductImageSetting|null
+     */
+    private function getProductImageSettingByImage(array $productImageSettings, int $imageId): ?ProductImageSetting
+    {
+        $productImageSettingsFiltered = array_filter(
+            $productImageSettings,
+            function (ProductImageSetting $productImageSetting) use ($imageId): bool {
+                return $productImageSetting->getProductImageId()->getValue() === $imageId;
+            }
+        );
+
+        return reset($productImageSettingsFiltered) ?: null;
+    }
+
+    /**
+     * @param ProductImageSetting[] $productImageSettings
+     * @param int $imageId
+     *
+     * @return array
+     */
+    private function extractShopsToAddImageTo(array $productImageSettings, int $imageId): array
+    {
+        $productImageSetting = $this->getProductImageSettingByImage($productImageSettings, $imageId);
+        $shopsToAddImageTo = [];
+        if ($productImageSetting) {
+            $shopsToAddImageTo = array_map(
+                static function (ShopId $shopId): int {
+                    return $shopId->getValue();
+                },
+                $productImageSetting->getShopIds()
+            );
+        }
+
+        return $shopsToAddImageTo;
+    }
+
+    /**
+     * @param array $shopIdsAssociatedToProduct
+     * @param array $shopsToAddImageTo
+     * @param \Image $image
+     *
+     * @return array
+     */
+    private function getShopsToRemoveImageFrom(array $shopIdsAssociatedToProduct, array $shopsToAddImageTo, \Image $image): array
+    {
+        $shopsToRemoveImageFrom = array_diff($shopIdsAssociatedToProduct, $shopsToAddImageTo);
+        $shopIdsAssociatedToImage = array_map(
+            function (ShopId $shopId): int {
+                return $shopId->getValue();
+            },
+            $this->productImageMultiShopRepository->getAssociatedShopIds(new ImageId($image->id))
+        );
+        $shopsToRemoveImageFrom = array_filter(
+            $shopsToRemoveImageFrom,
+            function (int $shopToRemoveImageFrom) use ($shopIdsAssociatedToImage): bool {
+                return in_array($shopToRemoveImageFrom, $shopIdsAssociatedToImage, true);
+            }
+        );
+
+        return $shopsToRemoveImageFrom;
+    }
+
+    /**
+     * @param int $productId
+     * @param \Image $image
+     * @param array $shopsToAddImageTo
+     *
+     * @return void
+     */
+    private function associateImageToShops(int $productId, \Image $image, array $shopsToAddImageTo): void
+    {
+        $image->id_product = $productId;
```

Comment:
This is still very weird that you need to handle this, especially it means that `$image->associateTo` can never be handled correctly alone, another develop will never know he needs to correctly associate `$image->id_product` beforehand and the association will fail

This is enough for now since it works, but at least a comment should be added to explain that this behaviour is not very stable or understood and deserves more research to understand what is failing at some point